Feature files with space in filename generate Illegal Character Exception (Junit5 Platform)

### 👓 What did you see?

When running test suite, if a name of a feature file contains a space character, program exits with :

```
Internal Error occurred.
org.junit.platform.commons.JUnitException: TestEngine with ID 'cucumber' failed to discover tests
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:160)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:134)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:80)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:110)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.IllegalArgumentException: Illegal character in path at index 49: classpath:/io/cucumber/junit/platform/engine/with space.feature
	at java.base/java.net.URI.create(URI.java:883)
	at io.cucumber.junit.platform.engine.FeatureOrigin.fromUri(FeatureOrigin.java:34)
	at io.cucumber.junit.platform.engine.FeatureResolver.createFeatureDescriptor(FeatureResolver.java:78)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.SortedOps$SizedRefSortingSink.end(SortedOps.java:357)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at io.cucumber.junit.platform.engine.FeatureResolver.resolveClasspathRoot(FeatureResolver.java:197)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at io.cucumber.junit.platform.engine.DiscoverySelectorResolver.resolve(DiscoverySelectorResolver.java:74)
	at io.cucumber.junit.platform.engine.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:47)
	at io.cucumber.junit.platform.engine.CucumberTestEngine.discover(CucumberTestEngine.java:43)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:152)
	... 13 more
Caused by: java.net.URISyntaxException: Illegal character in path at index 49: classpath:/io/cucumber/junit/platform/engine/with space.feature
	at java.base/java.net.URI$Parser.fail(URI.java:2913)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3084)
	at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3166)
	at java.base/java.net.URI$Parser.parse(URI.java:3114)
	at java.base/java.net.URI.<init>(URI.java:600)
	at java.base/java.net.URI.create(URI.java:881)
	... 29 more
```

### ✅ What did you expect to see?

Tests execution

### 📦 Which tool/library version are you using?

* JDK: 11 & 17
* Cucumber : 7.2.3 & 7.3.0-SNAPSHOT
* Junit: 5.8.2
* OS: Linux (Ubuntu)

### 🔬 How could we reproduce it?

Steps to reproduce the behavior:
1. Clone this repo
2. Create a file called 'with space.feature' into `junit-platform-engine/src/test/resources/io/cucumber/junit/platform/engine'
3. Run tests
4. See error

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -29,6 +29,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - [TestNG] Remove spurious Optional\[<Feature Name>] from test name ([#2488](https://github.com/cucumber/cucumber-jvm/pull/2488) M.P. Korstanje)
 * [BOM] Add missing dependencies to bill of materials ([#2496](https://github.com/cucumber/cucumber-jvm/pull/2496) M.P. Korstanje)
 * [Spring] Start and stop test context once per scenario ([#2517](https://github.com/cucumber/cucumber-jvm/pull/2517) M.P. Korstanje)
+* [JUnit Platform] Feature files with space in filename generate Illegal Character ([#2521](https://github.com/cucumber/cucumber-jvm/pull/2521) G. Fernandez)
 
 ## [7.2.3] (2022-01-13)
 

--- a/junit-platform-engine/src/main/java/io/cucumber/junit/platform/engine/FeatureOrigin.java
+++ b/junit-platform-engine/src/main/java/io/cucumber/junit/platform/engine/FeatureOrigin.java
@@ -31,7 +31,7 @@ abstract class FeatureOrigin {
             if (!uri.getSchemeSpecificPart().startsWith("/")) {
                 // ClasspathResourceSource.from expects all resources to start
                 // with a forward slash
-                uri = URI.create(CLASSPATH_SCHEME_PREFIX + "/" + uri.getSchemeSpecificPart());
+                uri = URI.create(CLASSPATH_SCHEME_PREFIX + "/" + uri.getRawSchemeSpecificPart());
             }
             ClasspathResourceSource source = ClasspathResourceSource.from(uri);
             return new ClasspathFeatureOrigin(source);

--- a/junit-platform-engine/src/test/java/io/cucumber/junit/platform/engine/DiscoverySelectorResolverTest.java
+++ b/junit-platform-engine/src/test/java/io/cucumber/junit/platform/engine/DiscoverySelectorResolverTest.java
@@ -79,6 +79,14 @@ class DiscoverySelectorResolverTest {
     }
 
     @Test
+    void resolveRequestWithClasspathResourceSelectorAndWithSpaceInFilename() {
+        DiscoverySelector resource = selectClasspathResource("io/cucumber/junit/platform/engine/with space.feature");
+        EngineDiscoveryRequest discoveryRequest = new SelectorRequest(resource);
+        resolver.resolveSelectors(discoveryRequest, testDescriptor);
+        assertEquals(1, testDescriptor.getChildren().size());
+    }
+
+    @Test
     void resolveRequestWithClasspathResourceSelectorAndFilePosition() {
         String feature = "io/cucumber/junit/platform/engine/rule.feature";
         FilePosition line = FilePosition.from(5);
@@ -120,7 +128,7 @@ class DiscoverySelectorResolverTest {
         DiscoverySelector resource = selectClasspathRoots(singleton(classpathRoot)).get(0);
         EngineDiscoveryRequest discoveryRequest = new SelectorRequest(resource);
         resolver.resolveSelectors(discoveryRequest, testDescriptor);
-        assertEquals(6, testDescriptor.getChildren().size());
+        assertEquals(7, testDescriptor.getChildren().size());
     }
 
     @Test
@@ -256,7 +264,7 @@ class DiscoverySelectorResolverTest {
         DiscoverySelector resource = selectDirectory("src/test/resources/io/cucumber/junit/platform/engine");
         EngineDiscoveryRequest discoveryRequest = new SelectorRequest(resource);
         resolver.resolveSelectors(discoveryRequest, testDescriptor);
-        assertEquals(5, testDescriptor.getChildren().size());
+        assertEquals(6, testDescriptor.getChildren().size());
     }
 
     @Test
@@ -264,7 +272,7 @@ class DiscoverySelectorResolverTest {
         DiscoverySelector resource = selectPackage("io.cucumber.junit.platform.engine");
         EngineDiscoveryRequest discoveryRequest = new SelectorRequest(resource);
         resolver.resolveSelectors(discoveryRequest, testDescriptor);
-        assertEquals(5, testDescriptor.getChildren().size());
+        assertEquals(6, testDescriptor.getChildren().size());
     }
 
     @Test
@@ -397,7 +405,7 @@ class DiscoverySelectorResolverTest {
         DiscoverySelector resource = selectClass(RunCucumberTest.class);
         EngineDiscoveryRequest discoveryRequest = new SelectorRequest(resource);
         resolver.resolveSelectors(discoveryRequest, testDescriptor);
-        assertEquals(5, testDescriptor.getChildren().size());
+        assertEquals(6, testDescriptor.getChildren().size());
     }
 
     @Test

--- a/junit-platform-engine/src/test/resources/io/cucumber/junit/platform/engine/with space.feature
+++ b/junit-platform-engine/src/test/resources/io/cucumber/junit/platform/engine/with space.feature
@@ -1,0 +1,1 @@
+Feature: A feature without any scenarios

--- a/junit-platform-engine/src/test/resources/io/cucumber/junit/platform/engine/with space.feature
+++ b/junit-platform-engine/src/test/resources/io/cucumber/junit/platform/engine/with space.feature
@@ -1,1 +1,6 @@
-Feature: A feature without any scenarios
+Feature: A feature with a single scenario inside a file with space in filename
+
+  Scenario: A single scenario
+    Given a single scenario
+    When it is executed
+    Then nothing else happens

